core/src/main/scala/shapeless/package.scala

```diff
@@ -114,28 +114,111 @@ package shapeless {
   @macrocompat.bundle
   class CachedImplicitMacros(val c: whitebox.Context) {
     import c.universe._
-
-    def dropLocal(nme: TermName): TermName = {
-      val LOCAL_SUFFIX_STRING = " "
-      val TermName(nmeString) = nme
-      if(nmeString endsWith LOCAL_SUFFIX_STRING)
-        TermName(nmeString.dropRight(LOCAL_SUFFIX_STRING.length))
-      else
-        nme
-    }
+    import scala.reflect.macros.TypecheckException
 
     def cachedImplicitImpl[T](implicit tTag: WeakTypeTag[T]): Tree = {
+      val casted = c.asInstanceOf[reflect.macros.runtime.Context]
+      val typer = casted.callsiteTyper
+      val global: casted.universe.type = casted.universe
+      val analyzer: global.analyzer.type = global.analyzer
+      val tCtx = typer.context
+      val owner = tCtx.owner
       val tTpe = weakTypeOf[T]
-      val owner = c.internal.enclosingOwner
-      val ownerNme = dropLocal(owner.name.toTermName)
-      val tpe = if(tTpe.typeSymbol.isParameter) owner.typeSignature else tTpe
-
-      q"""
-        {
-          def $ownerNme = ???
-          _root_.scala.Predef.implicitly[$tpe]
+      val application = casted.macroApplication
+      val tpe = {
+        if (tTpe.typeSymbol.isParameter) owner.tpe
+        else tTpe
+      }.asInstanceOf[global.Type]
+
+      // Run our own custom implicit search that isn't allowed to find
+      // the thing we are enclosed in
+      val sCtx = tCtx.makeImplicit(true)
+
+      // 2.12 changed the signature on a method we need to override. In order to allow us to define
+      // both versions below with the "override" modifier, we need an interface that contains both.
+      trait CompatImplicitSearch extends analyzer.ImplicitsContextErrors { self: analyzer.ImplicitSearch =>
```

Comment:
Could this be moved to macro-compat? Ideally we want shapeless to be written to just one version of the macro api/compiler internals and move all the version dependencies to macro-compat.